### PR TITLE
[stable/jenkins] Add Kubernetes plugin parameters

### PR DESCRIPTION
**Pull request by BryanHunt**
Signed-off-by: Bryan Hunt <bhunt@mac.com>

#### Is this a new chart

No

#### What this PR does / why we need it:

It add additional configuration parameters to the Kubernetes plugin

#### Which issue this PR fixes

  - fixes # 22212

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

File: stable/jenkins/README.md
Comment (wmcdona89) [inline stable/jenkins/README.md:112]: are you defining your agents in the chart via `agent` and `additionalAgents`?
all `additionalAgents` inherit from `agent` when the chart is deployed

that said, there can be value in setting Defaults Provider Template to ensure configuration changes made via the Jenkins console will be inherited as well

File: stable/jenkins/README.md
Comment (wmcdona89) [inline stable/jenkins/README.md:116]: this does seem like the right place for all these values as they're used for the top-level configuration of the kubernetes plugin while `agent` values are used to configure pod templates. the `master.slave*` names could stand to be refactored down the road...perhaps `master.k8s.*`.

File: stable/jenkins/README.md
Comment (wmcdona89) [inline stable/jenkins/README.md]: could `master.slaveJenkinsUrl` be used to set JENKINS_URL or is JENKINS_URL always a unique value?

issue #22219 references JENKINS_URL

File: stable/jenkins/values.yaml
Comment (wmcdona89) [inline stable/jenkins/values.yaml:194]: it seems like the defaults for `slaveConnectTimeout` and `slaveReadTimeout` should remain at 0

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml]: please also add `defaultsProviderTemplate` to the JCasC [here](https://github.com/helm/charts/blob/master/stable/jenkins/templates/_helpers.tpl#L91)

Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml]: should be `.Values.master.slaveDefaultsProviderTemplate`

File: stable/jenkins/values.yaml
Comment (wmcdona89) [reply stable/jenkins/values.yaml:194]: please include the other new configurable fields here as well along with useful comments. fields that don't have default values may be commented out.

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:71]: please also add `master.slaveJenkinsUrl` to the JCasC [here](https://github.com/helm/charts/blob/master/stable/jenkins/templates/_helpers.tpl#L93)

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:79]: please also add `master.slaveJenkinsTunnel` to the JCasC [here](https://github.com/helm/charts/blob/master/stable/jenkins/templates/_helpers.tpl#L92)

**Review by wmcdona89 — CHANGES_REQUESTED** [review]
thanks for the PR!

**Comment by wmcdona89** [comment]
Please also update the CHANGELOG

File: stable/jenkins/values.yaml
Comment (BryanHunt) [reply stable/jenkins/values.yaml:194]: The documentation for the Kubernetes plugin states that the minimum value for `slaveConnectTimeout` and `slaveReadTimeout` is 5 and 15 respectively.  When using the UI to configure those values, you cannot enter values less that the minimum.

If you want the to default to 0, I'll change it, but that seems to be an invalid value.

**Comment by BryanHunt** [comment]
I've made most of the changes requested.  I need to resolve the question on the JENKINS_URL environment variable and default provider template with a co-worker when they get in.

File: stable/jenkins/README.md
Comment (BryanHunt) [reply stable/jenkins/README.md:112]: This is what I got from one of my co-workers:

> defaultProviderTemplate sets podTemplate() step to use the podTemplate we provide in the k8s cloud configuration. Without setting defaultProviderTemplate the template is just there but not used unless we do podTemplate(inheritFrom:'default)

File: stable/jenkins/README.md
Comment (BryanHunt) [reply stable/jenkins/README.md]: We are good with setting the environment variable the same as `master.slaveJenkinsUrl`.  Shall I make that change?

File: stable/jenkins/README.md
Comment (wmcdona89) [reply stable/jenkins/README.md:112]: got it. you're only configuring `agent` in the chart and then defining pod templates in your pipelines which need to inherit from the pod template configured by `agent`. thus Defaults Provider Template needs to point to the pod template configured by `agent`.

File: stable/jenkins/README.md
Comment (wmcdona89) [reply stable/jenkins/README.md]: yes, please make the change. I think it would be confusing to have both `master.slaveJenkinsUrl` and `agent.jenkinsUrl`.

File: stable/jenkins/values.yaml
Comment (wmcdona89) [reply stable/jenkins/values.yaml:194]: It appears the Kubernetes plugin behaves differently when the Configuration As Code plugin is installed. Without the Configuration As Code plugin, the values get set to the minimums even though the chart configures them as 0. With the Configuration As Code plugin, a value of 0 is allowed and means no timeout. Let's leave the chart defaults as 0.

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:87]: please also add `master.slaveConnectTimeout` to the JCasC as `connectTimeout` [here](https://github.com/helm/charts/blob/master/stable/jenkins/templates/_helpers.tpl#L90)

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:88]: please also add `master.slaveReadTimeout`to the JCasC as `readTimeout` [here](https://github.com/helm/charts/blob/master/stable/jenkins/templates/_helpers.tpl#L101)

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl]: let's make this an `else if`

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl]: let's make this an `else if`

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml]: let's make this an `else if`

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml]: let's make this an `else if`

**Review by wmcdona89 — CHANGES_REQUESTED** [review]

**Comment by BryanHunt** [comment]
I think I've got all the requested changes in.  I still need to test locally and resolve the changelog conflict.

**Comment by BryanHunt** [comment]
My co-worker has a couple of issues that need to be sorted out.  Please don't merge this just yet.

**Comment by BryanHunt** [comment]
We are good to merge with my last fix if y'all are.

File: stable/jenkins/CHANGELOG.md
Comment (wmcdona89) [inline stable/jenkins/CHANGELOG.md:26]: looks like you missed committing your change description

File: stable/jenkins/values.yaml
Comment (wmcdona89) [inline stable/jenkins/values.yaml]: let's also add `master.slaveJenkinsUrl` and `master.slaveJenkinsTunnel` here so this file and the README are in sync

Comment (wmcdona89) [reply stable/jenkins/values.yaml]: oops...I discovered I was using an old version of the kubernetes plugin when looking into the minimum timeout values. looks like the minimums were established in [v1.21.5](https://github.com/jenkinsci/kubernetes-plugin/releases/tag/kubernetes-1.21.5)

so yeah...these should be 5 and 15 as you proposed

File: stable/jenkins/Chart.yaml
Comment (wmcdona89) [inline stable/jenkins/Chart.yaml]: let's make this a minor version bump

**Review by wmcdona89 — CHANGES_REQUESTED** [review]

File: stable/jenkins/CHANGELOG.md
Comment (BryanHunt) [reply stable/jenkins/CHANGELOG.md:26]: I used GitHub to resolve the merge conflict.  What do I need to do for this?

Comment (wmcdona89) [reply stable/jenkins/CHANGELOG.md:26]: this is currently just copy / paste of the verbiage from v1.18.1
it needs to be something like...

## 1.20.0

Add the following options for configuring the Kubernetes plugin.
- `master.slaveDefaultsProviderTemplate`
- `master.slaveJenkinsUrl`
- `master.slaveJenkinsTunnel`
- `master.slaveConnectTimeout`
-  `master.slaveReadTimeout`
```

File: stable/jenkins/Chart.yaml
Comment (wmcdona89) [inline stable/jenkins/Chart.yaml]: v1.19.0 was just merged. please rebase and bump to 1.20.0

**Review by wmcdona89 — CHANGES_REQUESTED** [review]

**Comment by BryanHunt** [comment]
Did something go wrong with the rebase?  Why are there so many commits?  Sorry, I'm not a git expert :(

**Comment by wmcdona89** [comment]
what you pushed looks more like a merge than a rebase. are you aware of `git push --force-with-lease`?

first ensure your fork is up to date. see [Syncing a fork](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) or [Syncing your fork to the original repository via the browser](https://github.com/KirstieJane/STEMMRoleModels/wiki/Syncing-your-fork-to-the-original-repository-via-the-browser).

then try the following...
```bash
git fetch 
git rebase -Xtheirs origin/master
# if the previous CHANGELOG entry was deleted, replace it and commit
git push --force-with-lease
```
[this article](https://demisx.github.io/git/rebase/2015/07/02/git-rebase-keep-my-branch-changes.html) explains `-Xtheirs`

**Comment by BryanHunt** [comment]
I'm getting error messages about cherry picking and could not apply a commit.  I'm going to close this and start over.  That will be like 10x faster than trying to fix whatever git did.

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl:95]: to allow templating this will be `jenkinsUrl: "{{ tpl .Values.master.slaveJenkinsUrl . }}"`

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl:103]: to allow templating this will be `jenkinsTunnel: "{{ tpl .Values.master.slaveJenkinsTunnel . }}"`

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl:288]: to allow templating this will be `<value>{{ tpl .Values.master.slaveJenkinsUrl . }}</value>`

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:71]: to allow templating this will be `<jenkinsUrl>{{ tpl .Values.master.slaveJenkinsUrl . }}</jenkinsUrl>`

File: stable/jenkins/templates/config.yaml
Comment (wmcdona89) [inline stable/jenkins/templates/config.yaml:79]: to allow templating this will be `<jenkinsTunnel>{{ tpl .Values.master.slaveJenkinsTunnel . }}</jenkinsTunnel>`

File: stable/jenkins/templates/_helpers.tpl
Comment (wmcdona89) [inline stable/jenkins/templates/_helpers.tpl:168]: to allow templating this will be `value: {{ tpl .Values.master.slaveJenkinsUrl . }}`

**Review by wmcdona89 — CHANGES_REQUESTED** [review]
it just occurred to me that `master.slaveJenkinsUrl` and `master.slaveJenkinsTunnel` should allow templating. if you don't mind making the updates I'd appreciate it. I believe there are 6 spots to update.

**Comment by wmcdona89** [comment]
> I'm getting error messages about cherry picking and could not apply a commit. I'm going to close this and start over. That will be like 10x faster than trying to fix whatever git did.

@BryanHunt good idea. please link to #22276 for reference. and one more request [here](https://github.com/helm/charts/pull/22276#pullrequestreview-412237838) if you don't mind.

**Pull request closed without merging**
